### DISCUSSION TITLE
Chaining `ts-node/esm` with `esmock` loader

### DISCUSSION THREAD

**Discussion by cjroebuck** [TypeStrong/ts-node#1748] [General]
_This was originally a bug report but has been converted into a discussion. The original bug report is below._

When wanting to use esmock with ts-node/esm loaders, you'll need to use [cspotcode/multiloader](https://github.com/cspotcode/multiloader) to allow node to use more than one --loader.

[esmock currently explicitly checks for --loader=esmock to be in process.execArgv](https://github.com/iambumblehead/esmock/blob/master/src/esmockLoader.mjs#L7-L8), so this also needs to be passed to node as a command line argument. It's not currently possible to use NODE_OPTIONS to run esmock alongside multiloader, though there is an [issue for that](https://github.com/iambumblehead/esmock/issues/52).

To get esmock and ts-node/esm to play nicely together you can use something like:

node --loader=esmock --loader @cspotcode/multiloader/compose?esmock,ts-node/esm

note that esmock is specified twice, once on it's own --loader and again inside the multiloader comma separated list of loaders. As mentioned above, this is mainly to get around the check in esmock, and is eventually ignored by node in favour of the second --loader.

This trick along with many others has helped me to get some jest tests working with typescript, esmodules and module mocking after many hours of frustration/pulling my hair out. 

A repo demonstrating such a setup with jest, typescript, esm and mocks is at: 

https://github.com/cjroebuck/jest-typescript-esmodules-esmock

### Search Terms

ERR_INVALID_MODULE_SPECIFIER
jest-light-runner
typescript esm

### Expected Behavior

It should transpile and run the typescript file

### Actual Behavior

ERR_INVALID_MODULE_SPECIFIER when running the command

### Steps to reproduce the problem

See the repro below

`npm run test` gives  TypeError [ERR_INVALID_MODULE_SPECIFIER]: Invalid module "file:///path/to/ts-node-repros/src/__tests__/foo.ts"

`npm run test:working` runs the tests successfully

The only difference is the --runInBand option which is implemented like this in jest-light-runner:

https://github.com/nicolo-ribaudo/jest-light-runner/pull/20/commits/fb7032596dab8f1db826179e98f001497f59ead0

I think it's something to do with dynamic import and perhaps the ts-node/esm loader is not being invoked in this case or something?

### Minimal reproduction

https://github.com/cjroebuck/ts-node-repro

### Specifications

ts-node v10.7.0
node v16.14.2
compiler v4.6.4
tsconfig.json, if you're using one:
```
{
    "extends": "ts-node/node16/tsconfig.json",
    "ts-node": {
        "swc": true
    },
    "compilerOptions": {
        "module": "ESNext"
    }
}
```
* Operating system and version: mac os monterey 12.0.1 (Apple M1 Max)

**Comment by cspotcode** [comment]
Might be a mistake in the way you're passing --loader?

**Comment by cjroebuck** [comment]
Yes, it seems that [esmock explicitly checks for --loader=esmock in argv](https://github.com/iambumblehead/esmock/blob/master/src/esmockLoader.mjs#L7-L8), so it is not currently possible to use the `NODE_OPTIONS` env variable to enable esmock at the same time as ts-node/esm. 

I'm also using your multiloader package for this so thank you for that too :)

The way I got this working in the end, with esmock and ts-node/esm loaders, was to change the command from:

`NODE_OPTIONS='--experimental-specifier-resolution=node --loader @cspotcode/multiloader/compose?esmock,ts-node/esm' node --loader=esmock node_modules/.bin/jest --watch --runInBand --testPathPattern __tests__/*`

to:

`node --experimental-specifier-resolution=node --loader=esmock --loader @cspotcode/multiloader/compose?esmock,ts-node/esm  node_modules/.bin/jest --watch --runInBand --testPathPattern __tests__/*`

The first --loader=esmock keeps esmock happy, but I'm assuming just gets ignored by node, because if I switch the order of the --loader's then I'm back to the original error (ERR_INVALID_MODULE_SPECIFIER).

**Comment by cspotcode** [comment]
Very clever, I'm glad it's working.  I expect things may change as soon as node releases support for chaining multiple loaders.  They have already merged the PR, so I expect that will be available in the next node release.  Then you'll be able to pass `--loader` multiple times, and node will take care of chaining them together.

Is this a bug in esmock?  It should support `NODE_OPTIONS`.  If you agree, I recommend opening an issue with them.

Given that this is not a bug report nor a feature request, but is rather a useful how-to guide, I'm going to convert it to a Discussion.

**Reply by cspotcode** [reply]
@cjroebuck I've changed the title of this discussion to describe what you were trying to do.  You may also want to edit the post to make it more helpful to other people who come here trying to accomplish the same thing.  It is basically a how-to guide.

**Reply by cjroebuck** [reply]
@cspotcode Thanks just updated the post now

**Comment by iambumblehead** [comment]
@cjroebuck this is awesome and NODE_OPTIONS support was [added to esmock](https://github.com/iambumblehead/esmock/releases/tag/v1.7.7)

I tried making a little test similar to yours but didn't succeed. Probably I need to follow your example precisely in order for it work

**Reply by iambumblehead** [reply]
@cjroebuck I'm interested to in adding an example jest test to esmock, but I'm unable to get it working. This is the command I'm using so far and the test runs but mock definitions aren't applied,
```json
{
    "test-jest": "NODE_OPTIONS='--experimental-vm-modules --loader=ts-node/esm --loader=esmock' jest ./spec/jest/*js",
}
```

**Comment by iambumblehead** [comment]
for anyone lurking this discussion: ts-node/esm and esmock do work correctly [as seen here.][0] My current questions relate to the jest runner specifically.

[0]: https://github.com/iambumblehead/esmock/blob/master/package.json#L69
